### PR TITLE
Allow the S3 bucket name to be specified via AWS_S3_BUCKET

### PR DESCRIPTION
**Pull request by justinsb**

### CODE REVIEW

**Comment by a-robinson** [comment]
LGTM, but
1) How would a new user be expected to know that this is available as an option? Would it hurt to also add it to the list in config-default.sh?
2) You'll have to rebase this on top of #5377 to make it mergable

**Comment by justinsb** [comment]
Good point.  I think we should probably have an options.md file or something like that, but for now I've taken your suggestion and put it into config-default.sh

We don't want to actually set a default (non-empty) value though, because we want to generate a (pseudo) unique name unless the user takes responsibility for picking the bucket name (because bucket names are global across regions _and users_).

**Comment by justinsb** [comment]
I fixed the merge-conflict here (not a real conflict, I had just changed adjacent lines).

File: cluster/aws/util.sh [hunk 1]
```diff
@@ -212,30 +212,34 @@ function upload-server-tars() {
...
-  if ! aws s3 ls "s3://${staging_bucket}" > /dev/null 2>&1 ; then
-    echo "Creating ${staging_bucket}"
-    aws s3 mb "s3://${staging_bucket}"
+  if ! aws s3 ls "s3://${AWS_S3_BUCKET}" > /dev/null 2>&1 ; then
+    echo "Creating ${AWS_S3_BUCKET}"
+    aws s3 mb "s3://${AWS_S3_BUCKET}"
   fi
 
-  local -r staging_path="${staging_bucket}/devel"
+  aws s3api put-bucket-acl --bucket ${AWS_S3_BUCKET} --acl public-read
```
Comment (a-robinson) [inline cluster/aws/util.sh]: This would undo #5377

**Comment by justinsb** [comment]
Wow... I'm not doing very well here.  Sorry & nice catch.  Let me fix this one.

I'm going to work on one patch at a time.  Starting with this one!

**Comment by a-robinson** [comment]
Ah, sorry about the delay here. I looked at it right after I got an email about your comment only to see that it still wasn't mergable. Looking again now.

File: cluster/aws/util.sh [hunk 2]
```diff
@@ -212,30 +212,32 @@ function upload-server-tars() {
...
-    project_hash=$(echo -n "${USER} ${key}" | md5sum | awk '{ print $1 }')
+  if [[ -z ${AWS_S3_BUCKET-} ]]; then
+      local project_hash=
+      local key=$(aws configure get aws_access_key_id)
+      if which md5 > /dev/null 2>&1; then
+        project_hash=$(md5 -q -s "${USER} ${key}")
+      else
+        project_hash=$(echo -n "${USER} ${key}" | md5sum | awk '{ print $1 }')
+      fi
+      AWS_S3_BUCKET="kubernetes-staging-${project_hash}/"
```
Comment (a-robinson) [inline cluster/aws/util.sh]: I don't think you want the slash on the end of the bucket name here, do you?

**Comment by justinsb** [comment]
Thanks for the catch (again).  I removed the extra trailing slash!

**Pull request merged**